### PR TITLE
Add a Service Worker and an offline page

### PR DESCRIPTION
**Pull request by chrisfinazzo**
Based on the discussion in #149, I took another crack at this. As the aim of Minima is to show what unmodified Jekyll looks like (E.g, `jekyll start`), I cached files based on the contents of `lib/site_template`. 

For a 3.0 release (#145), this seems like a reasonable thing to include.

### CODE REVIEW

**Comment by pathawks** [comment]
I am not sure that this is in keeping with the spirit of having a _minimal_ theme.

**Comment by chrisfinazzo** [comment]
@pathawks,

It probably isn't (in a strict sense), but this seems like a lighter-weight solution compared to using a module like Workbox. Although I've used that on a small site and it does the job well, it would require people to deal with NPM. That isn't difficult, but it is a bit heavy in terms of dependencies.

As previously mentioned, this is probably something for 3.0, so I'm happy to wait on this for a while.

**Comment by DirtyF** [comment]
Thanks @chrisfinazzo I'd really much like if we could provide a better user experience and allow offline browsing for minima. Shouldn't we add a timestamp to the `staticCacheName` here?

**Comment by DirtyF** [comment]
@chrisfinazzo we don't want any `npm` dependency to stay _minimal_. Thanks for using the native API.

**Comment by DirtyF** [comment]
[Going offline](https://abookapart.com/products/going-offline) by @adactio is out today, let's read it and make sure we have a solid strategy here 😉

**Comment by chrisfinazzo** [comment]
@DirtyF I read it this  afternoon, so this was on my mind 😃.

To your point regarding a timestamp on the `CacheName`, if we update the version constant, that _should_ cause a new registration. To be honest, the section on caching is a bit dense, so I probably need to spend a bit more time thinking about what the right thing to do is.

**Comment by DirtyF** [comment]
@chrisfinazzo  I was thinking of something like `const staticCacheName = "version-{{site.time | date: '%Y%m%d%H%M%S'}}";`

**Comment by chrisfinazzo** [comment]
@DirtyF Ah, I see now. Will update.

**Comment by ashawley** [comment]
This would be a nice addition.  Thanks for working on adding it.

1. The `offline.html` is a useful template, but wouldn't be shipped with the gem-based theme.
2. The same for `sw.js`.  It would need to be located in `assets/js` or some such.
3. Maybe call the file `offline.js` instead of `sw.js`?
4. If the service worker is added on every page, and the `sw.js` exists, but the `offline.html` does not, the service worker fetch will fail.  This is a Javascript console error.  Loading the site while offline will fail to load the offline page ⚡️, but the browser knows how to degrade gracefully to just a regular "This site can't be reached", so 👍.
5. It would be nice to have the offline page cache be configurable, but it seems like this should be a Jekyll plugin?

File: sw.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const staticCacheName = "version-{{site.time | date: '%Y%m%d%H%M%S'}}";
+
+addEventListener('install', installEvent => {
+     console.log('The service worker is installing...');
+     skipWaiting();
+     installEvent.waitUntil(
+       caches.open(staticCacheName)
+       .then( staticCache => {
+         return staticCache.addAll([
+           '/assets/css/style.css',
```
Comment (ashawley) [inline sw.js]: The name of the stylesheet was moved to `/assets/main.css`.

File: sw.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     })
+   );
+});
+
+
+addEventListener('fetch', fetchEvent => {
+     console.log('The service worker is listening.');
+     const request = fetchEvent.request;
+     fetchEvent.respondWith(
+       caches.match(staticCacheName)
```
Comment (ashawley) [inline sw.js]: This wasn't retrieving assets from the cache when offline, because it needs to be changed to `caches.match(request)`.

**Comment by chrisfinazzo** [comment]
@ashawley,

Thanks for the feedback. I have a branch in my local repo that has some of the changes you mentioned, I will review there and then add those to this PR.

File: sw.js (hunk 1 above)
Comment (chrisfinazzo) [reply sw.js]: Last I touched this, it had been [changed][] in order to match Pages' "house style". 

[changed]: https://github.com/jekyll/minima/pull/305/commits/7a64517c183fc4a823ebfee599806fd9ce509284

Comment (ashawley) [reply sw.js]: Ah, ok, so I had it backwards.  It got moved to `assets/css/style.css`.  That explains it.

**Comment by chrisfinazzo** [comment]
> 4. If the service worker is added on every page, and the `sw.js` exists, but the `offline.html` does not, the service worker fetch will fail.  This is a Javascript console error.  Loading the site while offline will fail to load the offline page ⚡️, but the browser knows how to degrade gracefully to just a regular "This site can't be reached", so 👍.


In GitHub Pages environments, the custom `404.html` is used. In terms of fallbacks, it's not much work to modify an `.htaccess` or `nginx.conf` to point to it - if your host doesn't do this behind the scenes already.

**Comment by ashawley** [comment]
I had suggested:

> 2. The same for `sw.js`.  It would need to be located in `assets/js` or some such.

According to [Google's primer on Service Workers](https://developers.google.com/web/fundamentals/primers/service-workers/) that wouldn't really work.

> One subtlety with the `register()` method is the location of the service worker file. You'll notice in this case that the service worker file is at the root of the domain. This means that the service worker's scope will be the entire origin. In other words, this service worker will receive fetch events for everything on this domain. If we register the service worker file at `/example/sw.js`, then the service worker would only see fetch events for pages whose URL starts with `/example/` (i.e. `/example/page1/`, `/example/page2/`).

It seems it would need to live at the root directory.

> 3. Maybe call the file `offline.js` instead of `sw.js`?

This name is cryptic, but I suppose it is the standard name for this kind of thing.

**Pull request closed without merging**